docs/architecture/adr-008.md

```diff
@@ -0,0 +1,191 @@
+
+### 2. Mapping Registration
+
+In this ADR, we do not intend to change the mapping registration process. Following the current design, the address specified in the network parameter `CronosAdmin` has the rights to update the mappings.
+
+### 3. Support EVM Token from/to Native Token Conversion
+
+#### A. CRC20/CRC21 Smart Contract Standards
+
+We want to leverage the existing CRC20/CRC21 token smart contract standards. Ideally no new method is introduced to support local EVM token to native token conversion.
+
+We can re-use bridge methods such as `send_to_ethereum(address recipient, uint amount, uint bridge_fee)` and `send_to_ibc(string memory recipient, uint amount)` methods to bridge tokens to different chains. If [ADR 007: Generic event formats for evm-hook actions](./adr-007.md) is accepted, this ADR is compatible as well.
+
+#### B. EVM Hooks
+
+EVM hooks for bridge applications are to be updated accordingly. Upon receiving the emitted bridge events, the EVM hook should:
+1. Unpack and validate the EVM events;
+2. Call `keeper.GetDenomByContract()` to validate and get the token source and native token denom of the EVM token smart contract address;
+3. (Optional) Validate if the native token denom is supported by the bridge^1;
+4. If the token is local EVM token, mint the corresponding native token and send to the user Cosmos address; Otherwise, send the native token from the contract address module account to the user Cosmos address;
+5. Interact with the bridge module and send the token to external chains;
+6. Emit Cosmos events as needed.
+
+^1 It is up to the design of each EVM hook to validate and restrict which the native denom can be bridged with the protocol. They can decide not to put any restriction to allow tokens transferring freely between differnt chains.
+
+#### C. Bridge Hooks
+
+Bridge hooks are to be updated according. Upon receiving bridge transfer in, the Bridge hook should, generally speaking, perform the followings:
+1. Unpack and validate bridge parameters from the transaction or event;
+2. If the token source is Cronos EVM, invoke `keeper.BurnVouchersToEvmCoins()`; Otherwise, invoke `keeper.ConvertVouchersToEvmCoins()`^2;
+3. Emit Cosmos events as needed.
+
+^2 BurnVouchersToEvmCoins will burn the native token, mint the local EVM tokens and send the token to the user EVM address. More details of the new keeper methods will be explained below.
+
+#### D. Native Denom Naming
+
+The denom naming of the native token with Cronos EVM as source is
+```
+evm0x{smartContractAddress}
+```
+
+The chain must avoid naming any native asset with the prefix `evm0x` in the future.
+
+### 4. Keeper Methods Changes
+
+### Method Updates
+
+#### GetDenomByContract
+
+```go
+GetDenomByContract(ctx sdk.Context, contract common.Address)
+    (source int, denom string, found bool)
+```
+
+`source` is added in the returned value to denote the source of the denom to smart contract address mapping
+
+##### GetDenomByContract
```

Comment:
Is this typo? should be GetContractByDenom?

Reply:
Thanks, updated!